[PyTorch][Fix] Pass the arguments of embedding as named arguments

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67574

When adding the optional params for sharded embedding op. Found that we cannot get these params from `__torch_function__` override. The reason is that we don't pass them via keyword arguments. So maybe we want to change them to kwargs?

Differential Revision: [D32039152](https://our.internmc.facebook.com/intern/diff/D32039152/)

cc @ezyang @gchanan

Commit:
[PyTorch][Fix] Pass the arguments of embedding as named arguments

When adding the optional params for sharded embedding op. Found that we cannot get these params from `__torch_function__` override. The reason is that we don't pass them via keyword arguments. So maybe we want to change them to kwargs?

Differential Revision: [D32039152](https://our.internmc.facebook.com/intern/diff/D32039152/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Fix] Pass the arguments of embedding as named arguments"


When adding the optional params for sharded embedding op. Found that we cannot get these params from `__torch_function__` override. The reason is that we don't pass them via keyword arguments. So maybe we want to change them to kwargs?

Differential Revision: [D32039152](https://our.internmc.facebook.com/intern/diff/D32039152/)

[ghstack-poisoned]

fduwjj commented:
@albanD sure. Let me keep both.

albanD commented:
What do you mean by both?

fduwjj commented:
@albanD I mean both positional arguments and kargs.

Commit:
Update on "[PyTorch][Fix] Pass the arguments of embedding as named arguments"


When adding the optional params for sharded embedding op. Found that we cannot get these params from `__torch_function__` override. The reason is that we don't pass them via keyword arguments. So maybe we want to change them to kwargs?

Differential Revision: [D32039152](https://our.internmc.facebook.com/intern/diff/D32039152/)

[ghstack-poisoned]

fduwjj commented:
Kept both positional arguments and kwargs. Added one more test case.

jbschlosser commented on torch/nn/functional.py:
```diff
@@ -2113,9 +2113,20 @@ def embedding(
-            embedding, (input, weight),
-            input, weight, padding_idx, max_norm, norm_type,
-            scale_grad_by_freq, sparse
+            embedding,
+            (input, weight),
+            input,
+            weight,
+            padding_idx,
+            max_norm,
+            norm_type,
+            scale_grad_by_freq,
+            sparse,
+            padding_idx=padding_idx,
+            max_norm=max_norm,
+            norm_type=norm_type,
+            scale_grad_by_freq=scale_grad_by_freq,
+            sparse=sparse,
```
I don't think passing both positional and kwarg forms is what we want - kwarg-only should be okay.

Unfortunately, whether args are passed positionally or as kwargs doesn't seem to be very consistent across functions in `torch.nn.functional` atm but ideally it would be.

fduwjj replied:
Do you feel that we should use kwarg for all ops? So basically in the shareded embedding op. If we don't pass in via kwarg. All the args will not be sent into `__torch_function__ `. Sharded embedding was added in this PR:  https://github.com/pytorch/pytorch/pull/66604. I am working on a PR which enables more input params for both Sharded Embedding and Embedding Bag. However, the parameters are only passed in for the embeddingBag case.

fduwjj replied:
Well, I can get these from positional args but this kindly make the code look hacky and bringing in implicit dependency. So I think we might want to pass in via kargs and use the key to get the parameter like max_norm, padding_idx, etc.

jbschlosser commented on test/test_nn.py:
```diff
@@ -4843,6 +4843,63 @@ def test_embedding_functional(self):
 
         self.assertEqual(res_old, res_F)
 
+        def test_embedding_with_kwargs(self):
```
It's not clear to me what this is testing - the logic here doesn't relate to `__torch_function__`. Imo you can just leave this test out.

fduwjj replied:
Thanks for reviewing and commenting. I also got the similar feeling as you. But since I was asked to add some tests for this PR, I tried to guess this is the place where the test may be. Or maybe you can let me know where should I put the test for this. My next PR around sharded embedding will cover this anyway.

fduwjj replied:
Here is PR I mentioned: https://github.com/pytorch/pytorch/pull/67799. Because of this one, I need to change it to kwargs. We can get it from positional args but that brings in implicit dependency. So I guess kwargs are better to go.

jbschlosser reviewed (commented):
Do you mind going into a bit more detail on the original problem you were running into? I'm interested in how receiving the args positionally into `__torch_function__` vs. receiving them as kwargs was blocking you. @fduwjj

fduwjj commented:
Let me know if you need to provide more context around the issue I mentioned above.  @jbschlosser

Commit:
Update on "[PyTorch][Fix] Pass the arguments of embedding as named arguments"


When adding the optional params for sharded embedding op. Found that we cannot get these params from `__torch_function__` override. The reason is that we don't pass them via keyword arguments. So maybe we want to change them to kwargs?

Differential Revision: [D32039152](https://our.internmc.facebook.com/intern/diff/D32039152/)

[ghstack-poisoned]

jbschlosser reviewed (approved):
I'm good with this since it matches what other functions (e.g. `embedding_bag`) do, but it is bc-breaking for those using `__torch_function__` with `embedding`. Marked the PR as such.

Commit:
Update on "[PyTorch][Fix] Pass the arguments of embedding as named arguments"


When adding the optional params for sharded embedding op. Found that we cannot get these params from `__torch_function__` override. The reason is that we don't pass them via keyword arguments. So maybe we want to change them to kwargs?

Differential Revision: [D32039152](https://our.internmc.facebook.com/intern/diff/D32039152/)

cc ezyang gchanan

[ghstack-poisoned]

albanD reviewed (approved):
sgtm

Commit:
Update on "[PyTorch][Fix] Pass the arguments of embedding as named arguments"


When adding the optional params for sharded embedding op. Found that we cannot get these params from `__torch_function__` override. The reason is that we don't pass them via keyword arguments. So maybe we want to change them to kwargs?

Differential Revision: [D32039152](https://our.internmc.facebook.com/intern/diff/D32039152/)

cc ezyang gchanan

[ghstack-poisoned]
